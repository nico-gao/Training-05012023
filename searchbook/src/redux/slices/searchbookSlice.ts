import { createSlice, PayloadAction, createAsyncThunk } from "@reduxjs/toolkit";
import { AppDispatch, RootState } from "../store";

export interface IBookItem {
    id: string;
    volumeInfo: {
        description: string | undefined;
        authors: string[] | undefined;
        title: string | undefined;
        publishedDate: string | undefined;
        imageLinks: { smallThumbnail: string | undefined; thumbnail: string | undefined } | undefined;
    };
}

interface SearchbookState {
    books: IBookItem[];
    keyword: string;
    isLoading: boolean;
}

const initialState: SearchbookState = {
    books: [],
    keyword: "",
    isLoading: false,
};

export const search = createAsyncThunk<
    // Return type of the payload creator
    any,
    // First argument to the payload creator
    undefined,
    {
        // Optional fields for defining thunkApi field types
        dispatch: AppDispatch;
        state: RootState;
    }
>("searchbook/search", async (args, thunkAPI) => {
    const { keyword } = thunkAPI.getState().searchbookSlice;
    const result = await fetch(
        `https://www.googleapis.com/books/v1/volumes?q=${keyword}&startIndex=0&maxResults=20`
    );
    const res = await result.json();
    console.log("res", res);
    return res.items;
});

const searchbookSlice = createSlice({
    name: "searchbook",
    initialState,
    reducers: {
        updateKeyword: (state, action) => {
            state.keyword = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(search.pending, (state, action) => {
                state.isLoading = true;
            })
            .addCase(search.fulfilled, (state, action) => {
                state.isLoading = false;
                state.books = action.payload;
                state.keyword = "";
            })
            .addCase(search.rejected, (state, action) => {
                console.log("err", action.error.message);
                state.isLoading = false;
                //show alert
            });
    },
});

export const { updateKeyword } = searchbookSlice.actions;

export default searchbookSlice.reducer;
