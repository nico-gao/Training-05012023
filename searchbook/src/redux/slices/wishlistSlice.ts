import { createSlice, PayloadAction, createAsyncThunk } from "@reduxjs/toolkit";
import { AppDispatch, RootState } from "../store";


interface BookItem {}

interface WishlistState {
    books: BookItem[];
}
const initialState: WishlistState = {
    books: [],
};



const wishlistSlice = createSlice({
    name: "wishlist",
    initialState,
    reducers: {
        
    },
    extraReducers: {},
});

export const {  } = wishlistSlice.actions;

export default wishlistSlice.reducer;