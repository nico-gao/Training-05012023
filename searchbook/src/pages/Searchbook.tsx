import React from "react";
import SearchBar from "../components/SearchBar";
import Booklist from "../components/Booklist";
import {useSelector} from "react-redux"
import { RootState } from "../redux/store";
import { IBookItem } from "../redux/slices/searchbookSlice";

const Searchbook = () => {
    const books = useSelector<RootState, IBookItem[]>(state=> state.searchbookSlice.books)
    return (
        <div>
            <SearchBar />
            <Booklist books={books}/>
        </div>
    );
};

export default Searchbook;
