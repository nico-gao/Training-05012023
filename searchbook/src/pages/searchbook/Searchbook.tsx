import React from "react";
import SearchBar from "../../components/SearchBar";
import Booklist from "../../components/Booklist";
import {useSelector} from "react-redux"
import { RootState } from "../../redux/store";
import { IBookItem } from "../../redux/slices/searchbookSlice";
/* 
    startIndex: currentPage * itemsPerPage
    maxResult: 20

    totalItems: 101;
    

    current page
    items per page
    total pages: Math.ceil(totalItem / itemsPerPage)
*/


const Searchbook = () => {
    const books = useSelector<RootState, IBookItem[]>(state=> state.searchbookSlice.books)
    return (
        <div className="searchbook">
            <SearchBar />
            <Booklist books={books}/>
            
        </div>
    );
};

export default Searchbook;
