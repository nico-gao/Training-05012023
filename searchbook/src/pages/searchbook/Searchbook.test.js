import { render, screen, waitFor } from "@testing-library/react";
import Searchbook from "./Searchbook";
import { configureStore } from "@reduxjs/toolkit";

import searchbookSlice from "../../redux/slices/searchbookSlice";
import wishlistSlice from "../../redux/slices/wishlistSlice";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";

const createMockStore = (
    preloadedState = {
        searchbookSlice: {
            books: [],
            keyword: "",
            isLoading: false,
            itemsPerPage: 10,
            totalPages: 1,
            currentPage: 1,
        },
        wishlistSlice: {
            books: [],
        },
    }
) => {
    return configureStore({
        reducer: {
            searchbookSlice,
            wishlistSlice,
        },
        preloadedState,
    });
};

fetch = jest.fn();

describe("Searchbook page", () => {
    beforeEach(() => {
        fetch.mockImplementation(() => {
            return Promise.resolve({
                json: () =>
                    Promise.resolve({
                        items: [
                            {
                                id: 1,
                                volumeInfo: {
                                    title: "",
                                    authors: [],
                                    publishedDate: "",
                                    description: "",
                                    imageLinks: {
                                        thumbnail: "",
                                        smallThumbnail: "",
                                    },
                                },
                            },
                            {
                                id: 2,
                                volumeInfo: {
                                    title: "",
                                    authors: [],
                                    publishedDate: "",
                                    description: "",
                                    imageLinks: {
                                        thumbnail: "",
                                        smallThumbnail: "",
                                    },
                                },
                            },
                        ],
                        kind: "",
                        totalItems: 2,
                    }),
            });
        });
    });

    test("snapshot", () => {
        const { asFragment } = render(
            <Provider store={createMockStore()}>
                <Searchbook />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
    test("should be able to type in the search bar", () => {
        render(
            <Provider store={createMockStore()}>
                <Searchbook />
            </Provider>
        );

        const inputEl = screen.getByRole("textbox");
        expect(inputEl).toBeInTheDocument();
        expect(inputEl).toHaveValue("");

        userEvent.type(inputEl, "hello");
        expect(inputEl).toHaveValue("hello");
    });

    test("results should show up after typing keywords and clicking the search button", async () => {
        render(
            <Provider store={createMockStore()}>
                <Searchbook />
            </Provider>
        );
        const inputEl = screen.getByRole("textbox");
        const searchBtnEl = screen.getByText("search");
        let liEls = screen.queryAllByRole("listitem"); // []
        expect(liEls).toHaveLength(0);
        expect(fetch).toBeCalledTimes(0);

        userEvent.type(inputEl, "hello");
        userEvent.click(searchBtnEl);

        liEls = await screen.findAllByRole("listitem");
        expect(liEls).toHaveLength(2);
        expect(fetch).toBeCalledTimes(1);
    });

    test("loader should show up when the data is loading", async () => {
        render(
            <Provider store={createMockStore()}>
                <Searchbook />
            </Provider>
        );

        let loaderEl = screen.queryByTestId("loader");
        expect(loaderEl).not.toBeInTheDocument();
        userEvent.type(screen.getByRole("textbox"), "hello");
        userEvent.click(screen.getByText("search"));

        loaderEl = screen.queryByTestId("loader");
        expect(loaderEl).toBeInTheDocument();
        // wait for the request to be fulfilled
        await waitFor(() => {
            loaderEl = screen.queryByTestId("loader");
            expect(loaderEl).not.toBeInTheDocument();
        });
    });
});
