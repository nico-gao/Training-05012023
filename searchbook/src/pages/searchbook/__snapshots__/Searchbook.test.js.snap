// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbook page snapshot 1`] = `
<DocumentFragment>
  <div
    class="searchbook"
  >
    <div
      class="search-bar"
    >
      <input
        value=""
      />
      <button>
        search
      </button>
    </div>
    <div
      class="book-list-container"
    >
      no result
    </div>
  </div>
</DocumentFragment>
`;
